R/preprocessing.R

```diff
@@ -177,3 +177,154 @@ default_fill_missing_obs <- function(data, obs, obs_column) {
+        ]
+      }
+    }
+    ## fill any missing columns
+    reported_cases_grid <- data.table::merge.data.table(
+      reported_cases, reported_cases_grid,
+      by = "date", all.y = TRUE
+    )
+  }
+  return(reported_cases[])
+}
+
+##' Add breakpoints to certain dates in a data set.
+##'
+##' @param dates A vector of dates to use as breakpoints.
+##' @inheritParams estimate_infections
+##' @return A data.table with `breakpoint` set to 1 on each of the specified
+##'   dates.
+##' @export
+##' @importFrom data.table setDT
+##' @examples
+##' reported_cases <- add_breakpoints(example_confirmed, as.Date("2020-03-26"))
+add_breakpoints <- function(data, dates = as.Date(character(0))) {
+  assert_data_frame(data)
+  assert_names(colnames(data), must.include = "date")
+  assert_date(dates)
+  assert_date(data$date, any.missing = FALSE)
+  reported_cases <- data.table::setDT(data)
+  if (is.null(reported_cases$breakpoint)) {
+    reported_cases$breakpoint <- 0
+  }
+  missing_dates <- setdiff(dates, data$date)
+  if (length(missing_dates) > 0) {
+    cli_abort("Breakpoint date{?s} not found in data: {.var {missing_dates}}")
+  }
+  reported_cases[date %in% dates, breakpoint := 1]
+  return(reported_cases)
+}
+
+##' Filter leading zeros from a data set.
+##'
+##' @inheritParams estimate_infections
+##' @inheritParams fill_missing
+##' @return A data.table with leading zeros removed.
+##' @export
+##' @importFrom data.table setDT
+##' @examples
+##' cases <- data.table(
+##'   date = as.Date("2020-01-01") + 0:10,
+##'   confirm = c(0, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9)
+##' )
+##' filter_leading_zeros(cases)
+filter_leading_zeros <- function(data, obs_column = "confirm", by = NULL) {
+  assert_data_frame(data)
+  assert_character(obs_column)
+  assert_character(by, null.ok = TRUE)
+  assert_names(
+    colnames(data),
+    must.include = c("date", by, obs_column)
+  )
+  reported_cases <- data.table::setDT(data)
+  reported_cases <- reported_cases[order(date)][
+    date >= min(date[get(obs_column)[!is.na(get(obs_column))] > 0])
+  ]
+  return(reported_cases[])
+}
+
+##' Converts zero case counts to NA (missing) if the 7-day average is above a
+##' threshold.
+##'
+##' This function aims to detect spurious zeroes by comparing the 7-day average
+##' of the case counts to a threshold. If the 7-day average is above the
+##' threshold, the zero case count is replaced with NA.
+##'
+##' @param threshold Numeric, defaults to Inf. Indicates if detected zero cases
+##'   are meaningful by using a threshold number of cases based on the 7-day
+##'   average. If the average is above this threshold at the time of a zero
+##'   observation count then the zero is replaced with a missing (`NA`) count
+##'   and thus ignored in the likelihood.
+##'
+##' @inheritParams estimate_infections
+##' @inheritParams fill_missing
+##' @importFrom data.table setDT frollsum
+##' @return A data.table with the zero threshold applied.
+##' @author Sebastian Funk
```

Comment:
```suggestion
```

Reply:
arg sorry I thought I'd stopped emacs from doing that but clearly I haven't